app/src/main/java/eu/kanade/domain/source/interactor/GetSourceFilterData.kt

```diff
@@ -0,0 +1,54 @@
+package eu.kanade.domain.source.interactor
+
+import eu.kanade.domain.source.model.Source
+import eu.kanade.domain.source.repository.SourceRepository
+import eu.kanade.tachiyomi.data.preference.PreferencesHelper
+import eu.kanade.tachiyomi.ui.browse.source.FilterUiModel
+import eu.kanade.tachiyomi.util.system.LocaleHelper
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.combine
+
+class GetSourceFilterData(
+    private val repository: SourceRepository,
+    private val preferences: PreferencesHelper,
+) {
+
+    fun subscribe(): Flow<List<FilterUiModel>> {
+        return combine(
+            preferences.enabledLanguages().asFlow(),
+            preferences.disabledSources().asFlow(),
+            repository.getOnlineSources()
+        ) { enabledLanguage, disabledSource, onlineSources ->
+            val sortedSources = onlineSources.sortedWith(
+                compareBy<Source> { it.id.toString() in disabledSource }
+                    .thenBy(String.CASE_INSENSITIVE_ORDER) { it.name }
+            )
+
+            val sortedSourcesByLang = sortedSources.groupBy { it.lang }
+                .toSortedMap(
+                    compareBy(
+                        { it !in enabledLanguage },
+                        { LocaleHelper.getDisplayName(it) }
+                    )
+                )
+
+            sortedSourcesByLang.flatMap {
+                val isLangEnabled = it.key in enabledLanguage
+                val items = mutableListOf<FilterUiModel>()
+
+                items.add(FilterUiModel.Header(it.key, isLangEnabled))
+                if (isLangEnabled) {
+                    it.value.map { source ->
+                        FilterUiModel.Item(
+                            source,
+                            source.id.toString() !in disabledSource
+                        )
+                    }.also { models ->
+                        items.addAll(models)
+                    }
+                }
+                items
+            }
```

Comment:
```suggestion
                val isLangEnabled = it.key in enabledLanguage
                val items = listOf(FilterUiModel.Header(it.key, isLangEnabled))
                if (isLangEnabled.not()) return@flatMap items
                items + it.value.map { source ->
                    FilterUiModel.Item(
                        source,
                        source.id.toString() !in disabledSource
                    )
                }
```
I would also move the flatmap to presentation layer